File: firebase-auth/src/test/java/com/androidhuman/rxfirebase2/auth/RxFirebaseUserTest.java
package com.androidhuman.rxfirebase2.auth;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.GetTokenResult;
import com.google.firebase.auth.UserProfileChangeRequest;

import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import io.reactivex.observers.TestObserver;

import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class RxFirebaseUserTest {

    @Mock
    FirebaseUser mockFirebaseUser;

    @Mock
    GetTokenResult mockGetTokenResult;

    @Mock
    AuthCredential mockAuthCredential;

    @Mock
    AuthResult mockAuthResult;

    @Mock
    UserProfileChangeRequest mockProfileChangeRequest;

    @Mock
    Task<Void> mockVoidTaskResult;

    @Mock
    Task<GetTokenResult> mockGetTokenTaskResult;

    @Mock
    Task<AuthResult> mockAuthTaskResult;

    private ArgumentCaptor<OnCompleteListener> onComplete;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);

        onComplete = ArgumentCaptor.forClass(OnCompleteListener.class);
    }

    @Test
    public void testDelete() {
        mockSuccessfulVoidResult();
        when(mockFirebaseUser.delete())
                .thenReturn(mockVoidTaskResult);

        TestObserver obs = TestObserver.create();

        RxFirebaseUser.delete(mockFirebaseUser)
                .subscribe(obs);

        callOnComplete(mockVoidTaskResult);
        obs.dispose();

        obs.assertComplete();
    }

    @Test
    public void testDelete_notSuccessful() {
        mockNotSuccessfulVoidResult(new IllegalStateException());
        when(mockFirebaseUser.delete())
                .thenReturn(mockVoidTaskResult);

        TestObserver obs = TestObserver.create();

        RxFirebaseUser.delete(mockFirebaseUser)
                .subscribe(obs);

        callOnComplete(mockVoidTaskResult);
        obs.dispose();

        obs.assertError(IllegalStateException.class);
    }

    @Test
    public void testGetToken() {
        mockSuccessfulTokenResult("token");
        //noinspection deprecation
        when(mockFirebaseUser.getToken(true))
                .thenReturn(mockGetTokenTaskResult);

        TestObserver<String> obs = TestObserver.create();

        //noinspection deprecation
        RxFirebaseUser.getToken(mockFirebaseUser, true)
                .subscribe(obs);

        callOnComplete(mockGetTokenTaskResult);
        obs.dispose();

        // Ensure no more values are emitted after unsubscribe
        callOnComplete(mockGetTokenTaskResult);

        obs.assertComplete();
        obs.assertValue("token");
    }

    @Test
    public void testGetToken_notSuccessful() {
        mockNotSuccessfulTokenResult(new IllegalStateException());
        //noinspection deprecation
        when(mockFirebaseUser.getToken(true))
                .thenReturn(mockGetTokenTaskResult);

        TestObserver<String> obs = TestObserver.create();

        //noinspection deprecation
        RxFirebaseUser.getToken(mockFirebaseUser, true)
                .subscribe(obs);

        callOnComplete(mockGetTokenTaskResult);
        obs.dispose();

        // Ensure no more values are emitted after unsubscribe
        callOnComplete(mockGetTokenTaskResult);

        obs.assertError(IllegalStateException.class);
    }

    @Test
    public void testGetIdToken() {
        mockSuccessfulTokenResult("token");
        when(mockFirebaseUser.getIdToken(true))
                .thenReturn(mockGetTokenTaskResult);

        TestObserver<String> obs = TestObserver.create();

        RxFirebaseUser.getIdToken(mockFirebaseUser, true)
                .subscribe(obs);

        callOnComplete(mockGetTokenTaskResult);
        obs.dispose();

        // Ensure no more values are emitted after unsubscribe
        callOnComplete(mockGetTokenTaskResult);

        obs.assertComplete();
        obs.assertValue("token");
    }

    @Test
    public void testGetIdToken_notSuccessful() {
        mockNotSuccessfulTokenResult(new IllegalStateException());
        when(mockFirebaseUser.getIdToken(true))
                .thenReturn(mockGetTokenTaskResult);

        TestObserver<String> obs = TestObserver.create();

        RxFirebaseUser.getIdToken(mockFirebaseUser, true)
                .subscribe(obs);

        callOnComplete(mockGetTokenTaskResult);
        obs.dispose();

        // Ensure no more values are emitted after unsubscribe
        callOnComplete(mockGetTokenTaskResult);

        obs.assertError(IllegalStateException.class);
    }

    @Test
    public void testLinkWithCredential() {
        mockSuccessfulAuthResult();
        when(mockFirebaseUser.linkWithCredential(mockAuthCredential))
                .thenReturn(mockAuthTaskResult);

        TestObserver<FirebaseUser> obs = TestObserver.create();

        RxFirebaseUser.linkWithCredential(mockFirebaseUser, mockAuthCredential)
                .subscribe(obs);

        callOnComplete(mockAuthTaskResult);
        obs.dispose();

        // Ensure no more values are emitted after unsubscribe
        callOnComplete(mockAuthTaskResult);

        obs.assertComplete();
        obs.assertValueCount(1);
    }

    @Test
    public void testLinkWithCredential_notSuccessful() {
        mockNotSuccessfulAuthResult(new IllegalStateException());
        when(mockFirebaseUser.linkWithCredential(mockAuthCredential))
                .thenReturn(mockAuthTaskResult);

        TestObserver<FirebaseUser> obs = TestObserver.create();

        RxFirebaseUser.linkWithCredential(mockFirebaseUser, mockAuthCredential)
                .subscribe(obs);

        callOnComplete(mockAuthTaskResult);
        obs.dispose();

        obs.assertError(IllegalStateException.class);
    }

    @Test
    public void testReauthenticate() {
        mockSuccessfulVoidResult();
        when(mockFirebaseUser.reauthenticate(mockAuthCredential))
                .thenReturn(mockVoidTaskResult);

        TestObserver obs = TestObserver.create();

        RxFirebaseUser.reauthenticate(mockFirebaseUser, mockAuthCredential)
                .subscribe(obs);

        callOnComplete(mockVoidTaskResult);
        obs.dispose();

        // Ensure no more values are emitted after unsubscribe
        callOnComplete(mockVoidTaskResult);

        obs.assertComplete();
    }

    @Test
    public void testReauthenticate_notSuccessful() {
        mockNotSuccessfulVoidResult(new IllegalStateException());
        when(mockFirebaseUser.reauthenticate(mockAuthCredential))
                .thenReturn(mockVoidTaskResult);

        TestObserver obs = TestObserver.create();

        RxFirebaseUser.reauthenticate(mockFirebaseUser, mockAuthCredential)
                .subscribe(obs);

        callOnComplete(mockVoidTaskResult);
        obs.dispose();

        obs.assertError(IllegalStateException.class);
    }

    @Test
    public void testReload() {
        mockSuccessfulVoidResult();
        when(mockFirebaseUser.reload())
                .thenReturn(mockVoidTaskResult);

        TestObserver obs = TestObserver.create();

        RxFirebaseUser.reload(mockFirebaseUser)
                .subscribe(obs);

        callOnComplete(mockVoidTaskResult);
        obs.dispose();

        obs.assertComplete();
    }

    @Test
    public void testReload_notSuccessful() {
        mockNotSuccessfulVoidResult(new IllegalStateException());
        when(mockFirebaseUser.reload())
                .thenReturn(mockVoidTaskResult);

        TestObserver obs = TestObserver.create();

        RxFirebaseUser.reload(mockFirebaseUser)
                .subscribe(obs);

        callOnComplete(mockVoidTaskResult);
        obs.dispose();

        obs.assertError(IllegalStateException.class);
    }

    @Test
    public void testSendEmailVerification() {
        mockSuccessfulVoidResult();
        when(mockFirebaseUser.sendEmailVerification())
                .thenReturn(mockVoidTaskResult);

        TestObserver obs = TestObserver.create();

        RxFirebaseUser.sendEmailVerification(mockFirebaseUser)
                .subscribe(obs);

        callOnComplete(mockVoidTaskResult);
        obs.dispose();

        obs.assertComplete();
    }

    @Test
    public void testSendEmailVerification_notSuccessful() {
        mockNotSuccessfulVoidResult(new IllegalStateException());
        when(mockFirebaseUser.sendEmailVerification())
                .thenReturn(mockVoidTaskResult);

        TestObserver obs = TestObserver.create();

        RxFirebaseUser.sendEmailVerification(mockFirebaseUser)
                .subscribe(obs);

        callOnComplete(mockVoidTaskResult);
        obs.dispose();

        obs.assertError(IllegalStateException.class);
    }

    @Test
    public void testUnlink() {
        mockSuccessfulAuthResult();
        when(mockFirebaseUser.unlink("provider"))
                .thenReturn(mockAuthTaskResult);

        TestObserver<FirebaseUser> obs = TestObserver.create();

        RxFirebaseUser.unlink(mockFirebaseUser, "provider")
                .subscribe(obs);

        callOnComplete(mockAuthTaskResult);
        obs.dispose();

        // Ensure no more values are emitted after unsubscribe
        callOnComplete(mockAuthTaskResult);

        obs.assertComplete();
        obs.assertValueCount(1);
    }

    @Test
    public void testUnlink_notSuccessful() {
        mockNotSuccessfulAuthResult(new IllegalStateException());
        when(mockFirebaseUser.unlink("provider"))
                .thenReturn(mockAuthTaskResult);

        TestObserver<FirebaseUser> obs = TestObserver.create();

        RxFirebaseUser.unlink(mockFirebaseUser, "provider")
                .subscribe(obs);

        callOnComplete(mockAuthTaskResult);
        obs.dispose();

        obs.assertError(IllegalStateException.class);
    }

    @Test
    public void testUpdateEmail() {
        mockSuccessfulVoidResult();
        when(mockFirebaseUser.updateEmail("foo@bar.com"))
                .thenReturn(mockVoidTaskResult);

        TestObserver obs = TestObserver.create();

        RxFirebaseUser.updateEmail(mockFirebaseUser, "foo@bar.com")
                .subscribe(obs);

        callOnComplete(mockVoidTaskResult);
        obs.dispose();

        obs.assertComplete();
    }

    @Test
    public void testUpdateEmail_notSuccessful() {
        mockNotSuccessfulVoidResult(new IllegalStateException());
        when(mockFirebaseUser.updateEmail("foo@bar.com"))
                .thenReturn(mockVoidTaskResult);

        TestObserver obs = TestObserver.create();

        RxFirebaseUser.updateEmail(mockFirebaseUser, "foo@bar.com")
                .subscribe(obs);

        callOnComplete(mockVoidTaskResult);
        obs.dispose();

        obs.assertError(IllegalStateException.class);
    }

    @Test
    public void testUpdatePassword() {
        mockSuccessfulVoidResult();
        when(mockFirebaseUser.updatePassword("password"))
                .thenReturn(mockVoidTaskResult);

        TestObserver obs = TestObserver.create();

        RxFirebaseUser.updatePassword(mockFirebaseUser, "password")
                .subscribe(obs);

        callOnComplete(mockVoidTaskResult);
        obs.dispose();

        obs.assertComplete();
    }

    @Test
    public void testUpdatePassword_notSuccessful() {
        mockNotSuccessfulVoidResult(new IllegalStateException());
        when(mockFirebaseUser.updatePassword("password"))
                .thenReturn(mockVoidTaskResult);

        TestObserver obs = TestObserver.create();

        RxFirebaseUser.updatePassword(mockFirebaseUser, "password")
                .subscribe(obs);

        callOnComplete(mockVoidTaskResult);
        obs.dispose();

        obs.assertError(IllegalStateException.class);
    }

    @Test
    public void testUpdateProfile() {
        mockSuccessfulVoidResult();
        when(mockFirebaseUser.updateProfile(mockProfileChangeRequest))
                .thenReturn(mockVoidTaskResult);

        TestObserver obs = TestObserver.create();

        RxFirebaseUser.updateProfile(mockFirebaseUser, mockProfileChangeRequest)
                .subscribe(obs);

        callOnComplete(mockVoidTaskResult);
        obs.dispose();

        obs.assertComplete();
    }

    @Test
    public void testUpdateProfile_notSuccessful() {
        mockNotSuccessfulVoidResult(new IllegalStateException());
        when(mockFirebaseUser.updateProfile(mockProfileChangeRequest))
                .thenReturn(mockVoidTaskResult);

        TestObserver obs = TestObserver.create();

        RxFirebaseUser.updateProfile(mockFirebaseUser, mockProfileChangeRequest)
                .subscribe(obs);

        callOnComplete(mockVoidTaskResult);
        obs.dispose();

        obs.assertError(IllegalStateException.class);
    }

    private void mockSuccessfulVoidResult() {
        mockVoidResult(true, null);
    }

    private void mockNotSuccessfulVoidResult(@NonNull Exception exception) {
        mockVoidResult(false, exception);
    }

    private void mockVoidResult(boolean success, @Nullable Exception exception) {
        when(mockVoidTaskResult.isSuccessful())
                .thenReturn(success);

        if (null != exception) {
            when(mockVoidTaskResult.getException())
                    .thenReturn(exception);
        }

        //noinspection unchecked
        when(mockVoidTaskResult.addOnCompleteListener(onComplete.capture()))
                .thenReturn(mockVoidTaskResult);
    }

    private void mockSuccessfulTokenResult(String token) {
        when(mockGetTokenResult.getToken())
                .thenReturn(token);

        when(mockGetTokenTaskResult.isSuccessful())
                .thenReturn(true);

        when(mockGetTokenTaskResult.getResult())
                .thenReturn(mockGetTokenResult);

        //noinspection unchecked
        when(mockGetTokenTaskResult.addOnCompleteListener(onComplete.capture()))
                .thenReturn(mockGetTokenTaskResult);
    }

    private void mockNotSuccessfulTokenResult(Exception exception) {
        when(mockGetTokenTaskResult.isSuccessful())
                .thenReturn(false);

        when(mockGetTokenTaskResult.getException())
                .thenReturn(exception);

        when(mockGetTokenTaskResult.getResult())
                .thenReturn(mockGetTokenResult);

        //noinspection unchecked
        when(mockGetTokenTaskResult.addOnCompleteListener(onComplete.capture()))
                .thenReturn(mockGetTokenTaskResult);
    }

    private void mockSuccessfulAuthResult() {
        when(mockAuthTaskResult.isSuccessful())
                .thenReturn(true);

        when(mockAuthResult.getUser())
                .thenReturn(mockFirebaseUser);

        when(mockAuthTaskResult.getResult())
                .thenReturn(mockAuthResult);

        //noinspection unchecked
        when(mockAuthTaskResult.addOnCompleteListener(onComplete.capture()))
                .thenReturn(mockAuthTaskResult);
    }

    private void mockNotSuccessfulAuthResult(Exception exception) {
        when(mockAuthTaskResult.isSuccessful())
                .thenReturn(false);

        when(mockAuthTaskResult.getException())
                .thenReturn(exception);

        when(mockAuthTaskResult.getResult())
                .thenReturn(mockAuthResult);

        //noinspection unchecked
        when(mockAuthTaskResult.addOnCompleteListener(onComplete.capture()))
                .thenReturn(mockAuthTaskResult);
    }

    @SuppressWarnings("unchecked")
    private void callOnComplete(Task<?> task) {
        verify(task)
                .addOnCompleteListener(onComplete.capture());
        onComplete.getValue().onComplete(task);
    }
}

File: firebase-auth/src/test/kotlin/com/androidhuman/rxfirebase2/auth/RxFirebaseUserTest.kt
package com.androidhuman.rxfirebase2.auth

import com.google.android.gms.tasks.OnCompleteListener
import com.google.android.gms.tasks.Task
import com.google.firebase.auth.AuthCredential
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.GetTokenResult
import com.google.firebase.auth.PhoneAuthCredential
import com.google.firebase.auth.UserProfileChangeRequest
import com.nhaarman.mockito_kotlin.argumentCaptor
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.times
import com.nhaarman.mockito_kotlin.verify
import com.nhaarman.mockito_kotlin.whenever
import io.reactivex.observers.TestObserver
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import org.mockito.MockitoAnnotations

@Suppress("NOTHING_TO_INLINE", "DEPRECATION")
class RxFirebaseUserTest {

    @Mock
    private lateinit var firebaseUser: FirebaseUser

    private val onAuthResultCompleteListener
            = argumentCaptor<OnCompleteListener<AuthResult>>()

    private val onGetTokenResultCompleteListener
            = argumentCaptor<OnCompleteListener<GetTokenResult>>()

    private val onVoidCompleteListener
            = argumentCaptor<OnCompleteListener<Void>>()

    @Before
    fun setup() {
        MockitoAnnotations.initMocks(this)
    }

    @Test
    fun delete() {
        val task = succeedVoidTask()

        whenever(firebaseUser.delete())
                .thenReturn(task)

        with(TestObserver.create<Any>()) {
            RxFirebaseUser.delete(firebaseUser)
                    .subscribe(this)

            // verify delete() has called
            verify(firebaseUser, times(1))
                    .delete()

            // verify addOnCompleteListener() has called
            task.verifyAddOnCompleteListenerCalled()

            // simulate the callback
            onVoidCompleteListener.lastValue.onComplete(task)

            assertNoErrors()
            assertComplete()

            dispose()
        }
    }

    @Test
    fun deleteNotSuccessful() {
        val task = failedTask<Void>(IllegalStateException())

        whenever(firebaseUser.delete())
                .thenReturn(task)

        with(TestObserver.create<Any>()) {
            RxFirebaseUser.delete(firebaseUser)
                    .subscribe(this)

            // verify delete() has called
            verify(firebaseUser, times(1))
                    .delete()

            // verify addOnCompleteListener() has called
            task.verifyAddOnCompleteListenerCalled()

            // simulate the callback
            onVoidCompleteListener.lastValue.onComplete(task)

            assertError(IllegalStateException::class.java)

            dispose()
        }
    }

    @Test
    fun getToken() {
        val task = succeedGetTokenResultTask("token")

        whenever(firebaseUser.getToken(true))
                .thenReturn(task)

        with(TestObserver.create<String>()) {
            RxFirebaseUser.getToken(firebaseUser, true)
                    .subscribe(this)

            // verify delete() has called
            verify(firebaseUser, times(1))
                    .getToken(true)

            // verify addOnCompleteListener() has called
            task.verifyAddOnCompleteListenerCalled()

            // simulate the callback
            onGetTokenResultCompleteListener.lastValue.onComplete(task)

            assertNoErrors()
            assertComplete()

            dispose()
        }
    }

    @Test
    fun getTokenNotSuccessful() {
        val task = failedTask<GetTokenResult>(IllegalStateException())

        whenever(firebaseUser.getToken(true))
                .thenReturn(task)

        with(TestObserver.create<String>()) {
            RxFirebaseUser.getToken(firebaseUser, true)
                    .subscribe(this)

            // verify delete() has called
            verify(firebaseUser, times(1))
                    .getToken(true)

            // verify addOnCompleteListener() has called
            task.verifyAddOnCompleteListenerCalled()

            // simulate the callback
            onGetTokenResultCompleteListener.lastValue.onComplete(task)

            assertError(IllegalStateException::class.java)

            dispose()
        }
    }

    @Test
    fun getIdToken() {
        val task = succeedGetTokenResultTask("token")

        whenever(firebaseUser.getIdToken(true))
                .thenReturn(task)

        with(TestObserver.create<String>()) {
            RxFirebaseUser.getIdToken(firebaseUser, true)
                    .subscribe(this)

            // verify delete() has called
            verify(firebaseUser, times(1))
                    .getIdToken(true)

            // verify addOnCompleteListener() has called
            task.verifyAddOnCompleteListenerCalled()

            // simulate the callback
            onGetTokenResultCompleteListener.lastValue.onComplete(task)

            assertNoErrors()
            assertComplete()

            dispose()
        }
    }

    @Test
    fun getIdTokenNotSuccessful() {
        val task = failedTask<GetTokenResult>(IllegalStateException())

        whenever(firebaseUser.getIdToken(true))
                .thenReturn(task)

        with(TestObserver.create<String>()) {
            RxFirebaseUser.getIdToken(firebaseUser, true)
                    .subscribe(this)

            // verify delete() has called
            verify(firebaseUser, times(1))
                    .getIdToken(true)

            // verify addOnCompleteListener() has called
            task.verifyAddOnCompleteListenerCalled()

            // simulate the callback
            onGetTokenResultCompleteListener.lastValue.onComplete(task)

            assertError(IllegalStateException::class.java)

            dispose()
        }
    }

    @Test
    fun linkWithCredential() {
        val credential = mock<AuthCredential>()
        val task = succeedAuthResultTask("foo@bar.com")

        whenever(firebaseUser.linkWithCredential(credential))
                .thenReturn(task)

        with(TestObserver.create<FirebaseUser>()) {
            RxFirebaseUser.linkWithCredential(firebaseUser, credential)
                    .subscribe(this)

            // verify linkWithCredential() has called
            verify(firebaseUser, times(1))
                    .linkWithCredential(credential)

            // verify addOnCompleteListener() has called
            task.verifyAddOnCompleteListenerCalled()

            // simulate the callback
            onAuthResultCompleteListener.lastValue.onComplete(task)

            assertNoErrors()
            assertComplete()

            dispose()
        }
    }

    @Test
    fun linkWithCredentialNotSuccessful() {
        val credential = mock<AuthCredential>()
        val task = failedTask<AuthResult>(IllegalStateException())

        whenever(firebaseUser.linkWithCredential(credential))
                .thenReturn(task)

        with(TestObserver.create<FirebaseUser>()) {
            RxFirebaseUser.linkWithCredential(firebaseUser, credential)
                    .subscribe(this)

            // verify linkWithCredential() has called
            verify(firebaseUser, times(1))
                    .linkWithCredential(credential)

            // verify addOnCompleteListener() has called
            task.verifyAddOnCompleteListenerCalled()

            // simulate the callback
            onAuthResultCompleteListener.lastValue.onComplete(task)

            assertError(IllegalStateException::class.java)

            dispose()
        }
    }

    @Test
    fun reauthenticate() {
        val credential = mock<AuthCredential>()
        val task = succeedVoidTask()

        whenever(firebaseUser.reauthenticate(credential))
                .thenReturn(task)

        with(TestObserver.create<Any>()) {
            RxFirebaseUser.reauthenticate(firebaseUser, credential)
                    .subscribe(this)

            // verify reauthenticate() has called
            verify(firebaseUser, times(1))
                    .reauthenticate(credential)

            // verify addOnCompleteListener() has called
            task.verifyAddOnCompleteListenerCalled()

            // simulate the callback
            onVoidCompleteListener.lastValue.onComplete(task)

            assertNoErrors()
            assertComplete()

            dispose()
        }
    }

    @Test
    fun reauthenticateNotSuccessful() {
        val credential = mock<AuthCredential>()
        val task = failedTask<Void>(IllegalStateException())

        whenever(firebaseUser.reauthenticate(credential))
                .thenReturn(task)

        with(TestObserver.create<Any>()) {
            RxFirebaseUser.reauthenticate(firebaseUser, credential)
                    .subscribe(this)

            // verify reauthenticate() has called
            verify(firebaseUser, times(1))
                    .reauthenticate(credential)

            // verify addOnCompleteListener() has called
            task.verifyAddOnCompleteListenerCalled()

            // simulate the callback
            onVoidCompleteListener.lastValue.onComplete(task)

            assertError(IllegalStateException::class.java)

            dispose()
        }
    }

    @Test
    fun reload() {
        val task = succeedVoidTask()

        whenever(firebaseUser.reload())
                .thenReturn(task)

        with(TestObserver.create<Any>()) {
            RxFirebaseUser.reload(firebaseUser)
                    .subscribe(this)

            // verify reload() has called
            verify(firebaseUser, times(1))
                    .reload()

            // verify addOnCompleteListener() has called
            task.verifyAddOnCompleteListenerCalled()

            // simulate the callback
            onVoidCompleteListener.lastValue.onComplete(task)

            assertNoErrors()
            assertComplete()

            dispose()
        }
    }

    @Test
    fun reloadNotSuccessful() {
        val task = failedTask<Void>(IllegalStateException())

        whenever(firebaseUser.reload())
                .thenReturn(task)

        with(TestObserver.create<Any>()) {
            RxFirebaseUser.reload(firebaseUser)
                    .subscribe(this)

            // verify reload() has called
            verify(firebaseUser, times(1))
                    .reload()

            // verify addOnCompleteListener() has called
            task.verifyAddOnCompleteListenerCalled()

            // simulate the callback
            onVoidCompleteListener.lastValue.onComplete(task)

            assertError(IllegalStateException::class.java)

            dispose()
        }
    }

    @Test
    fun sendEmailVerification() {
        val task = succeedVoidTask()

        whenever(firebaseUser.sendEmailVerification())
                .thenReturn(task)

        with(TestObserver.create<Any>()) {
            RxFirebaseUser.sendEmailVerification(firebaseUser)
                    .subscribe(this)

            // verify sendEmailVerification() has called
            verify(firebaseUser, times(1))
                    .sendEmailVerification()

            // verify addOnCompleteListener() has called
            task.verifyAddOnCompleteListenerCalled()

            // simulate the callback
            onVoidCompleteListener.lastValue.onComplete(task)

            assertNoErrors()
            assertComplete()

            dispose()
        }
    }

    @Test
    fun sendEmailVerificationNotSuccessful() {
        val task = failedTask<Void>(IllegalStateException())

        whenever(firebaseUser.sendEmailVerification())
                .thenReturn(task)

        with(TestObserver.create<Any>()) {
            RxFirebaseUser.sendEmailVerification(firebaseUser)
                    .subscribe(this)

            // verify sendEmailVerification() has called
            verify(firebaseUser, times(1))
                    .sendEmailVerification()

            // verify addOnCompleteListener() has called
            task.verifyAddOnCompleteListenerCalled()

            // simulate the callback
            onVoidCompleteListener.lastValue.onComplete(task)

            assertError(IllegalStateException::class.java)

            dispose()
        }
    }

    @Test
    fun unlink() {
        val task = succeedAuthResultTask("foo@bar.com")

        whenever(firebaseUser.unlink("provider"))
                .thenReturn(task)

        with(TestObserver.create<FirebaseUser>()) {
            RxFirebaseUser.unlink(firebaseUser, "provider")
                    .subscribe(this)

            // verify unlink() has called
            verify(firebaseUser, times(1))
                    .unlink("provider")

            // verify addOnCompleteListener() has called
            task.verifyAddOnCompleteListenerCalled()

            // simulate the callback
            onAuthResultCompleteListener.lastValue.onComplete(task)

            assertNoErrors()
            assertComplete()

            dispose()
        }
    }

    @Test
    fun unlinkNotSuccessful() {
        val task = failedTask<AuthResult>(IllegalStateException())

        whenever(firebaseUser.unlink("provider"))
                .thenReturn(task)

        with(TestObserver.create<FirebaseUser>()) {
            RxFirebaseUser.unlink(firebaseUser, "provider")
                    .subscribe(this)

            // verify unlink() has called
            verify(firebaseUser, times(1))
                    .unlink("provider")

            // verify addOnCompleteListener() has called
            task.verifyAddOnCompleteListenerCalled()

            // simulate the callback
            onAuthResultCompleteListener.lastValue.onComplete(task)

            assertError(IllegalStateException::class.java)

            dispose()
        }
    }

    @Test
    fun updateEmail() {
        val task = succeedVoidTask()

        whenever(firebaseUser.updateEmail("foo@bar.com"))
                .thenReturn(task)

        with(TestObserver.create<Any>()) {
            RxFirebaseUser.updateEmail(firebaseUser, "foo@bar.com")
                    .subscribe(this)

            // verify updateEmail() has called
            verify(firebaseUser, times(1))
                    .updateEmail("foo@bar.com")

            // verify addOnCompleteListener() has called
            task.verifyAddOnCompleteListenerCalled()

            // simulate the callback
            onVoidCompleteListener.lastValue.onComplete(task)

            assertNoErrors()
            assertComplete()

            dispose()
        }
    }

    @Test
    fun updateEmailNotSuccessful() {
        val task = failedTask<Void>(IllegalStateException())

        whenever(firebaseUser.updateEmail("foo@bar.com"))
                .thenReturn(task)

        with(TestObserver.create<Any>()) {
            RxFirebaseUser.updateEmail(firebaseUser, "foo@bar.com")
                    .subscribe(this)

            // verify updateEmail() has called
            verify(firebaseUser, times(1))
                    .updateEmail("foo@bar.com")

            // verify addOnCompleteListener() has called
            task.verifyAddOnCompleteListenerCalled()

            // simulate the callback
            onVoidCompleteListener.lastValue.onComplete(task)

            assertError(IllegalStateException::class.java)

            dispose()
        }
    }

    @Test
    fun updatePassword() {
        val task = succeedVoidTask()

        whenever(firebaseUser.updatePassword("password"))
                .thenReturn(task)

        with(TestObserver.create<Any>()) {
            RxFirebaseUser.updatePassword(firebaseUser, "password")
                    .subscribe(this)

            // verify updatePassword() has called
            verify(firebaseUser, times(1))
                    .updatePassword("password")

            // verify addOnCompleteListener() has called
            task.verifyAddOnCompleteListenerCalled()

            // simulate the callback
            onVoidCompleteListener.lastValue.onComplete(task)

            assertNoErrors()
            assertComplete()

            dispose()
        }
    }

    @Test
    fun updatePasswordNotSuccessful() {
        val task = failedTask<Void>(IllegalStateException())

        whenever(firebaseUser.updatePassword("password"))
                .thenReturn(task)

        with(TestObserver.create<Any>()) {
            RxFirebaseUser.updatePassword(firebaseUser, "password")
                    .subscribe(this)

            // verify updatePassword() has called
            verify(firebaseUser, times(1))
                    .updatePassword("password")

            // verify addOnCompleteListener() has called
            task.verifyAddOnCompleteListenerCalled()

            // simulate the callback
            onVoidCompleteListener.lastValue.onComplete(task)

            assertError(IllegalStateException::class.java)

            dispose()
        }
    }

    @Test
    fun updatePhoneNumber() {
        val task = succeedVoidTask()
        val credential = mock<PhoneAuthCredential>()

        whenever(firebaseUser.updatePhoneNumber(credential))
                .thenReturn(task)

        with(TestObserver.create<Any>()) {
            RxFirebaseUser.updatePhoneNumber(firebaseUser, credential)
                    .subscribe(this)

            // verify updatePhoneNumber() has called
            verify(firebaseUser, times(1))
                    .updatePhoneNumber(credential)

            // verify addOnCompleteListener() has called
            task.verifyAddOnCompleteListenerCalled()

            // simulate the callback
            onVoidCompleteListener.lastValue.onComplete(task)

            assertNoErrors()
            assertComplete()

            dispose()
        }
    }

    @Test
    fun updatePhoneNumberNotSuccessful() {
        val task = failedTask<Void>(IllegalStateException())
        val credential = mock<PhoneAuthCredential>()

        whenever(firebaseUser.updatePhoneNumber(credential))
                .thenReturn(task)

        with(TestObserver.create<Any>()) {
            RxFirebaseUser.updatePhoneNumber(firebaseUser, credential)
                    .subscribe(this)

            // verify updatePhoneNumber() has called
            verify(firebaseUser, times(1))
                    .updatePhoneNumber(credential)

            // verify addOnCompleteListener() has called
            task.verifyAddOnCompleteListenerCalled()

            // simulate the callback
            onVoidCompleteListener.lastValue.onComplete(task)

            assertError(IllegalStateException::class.java)

            dispose()
        }
    }

    @Test
    fun updateProfile() {
        val profile = mock<UserProfileChangeRequest>()
        val task = succeedVoidTask()

        whenever(firebaseUser.updateProfile(profile))
                .thenReturn(task)

        with(TestObserver.create<Any>()) {
            RxFirebaseUser.updateProfile(firebaseUser, profile)
                    .subscribe(this)

            // verify updateProfile() has called
            verify(firebaseUser, times(1))
                    .updateProfile(profile)

            // verify addOnCompleteListener() has called
            task.verifyAddOnCompleteListenerCalled()

            // simulate the callback
            onVoidCompleteListener.lastValue.onComplete(task)

            assertNoErrors()
            assertComplete()

            dispose()
        }
    }

    @Test
    fun updateProfileNotSuccessful() {
        val profile = mock<UserProfileChangeRequest>()
        val task = failedTask<Void>(IllegalStateException())

        whenever(firebaseUser.updateProfile(profile))
                .thenReturn(task)

        with(TestObserver.create<Any>()) {
            RxFirebaseUser.updateProfile(firebaseUser, profile)
                    .subscribe(this)

            // verify updateProfile() has called
            verify(firebaseUser, times(1))
                    .updateProfile(profile)

            // verify addOnCompleteListener() has called
            task.verifyAddOnCompleteListenerCalled()

            // simulate the callback
            onVoidCompleteListener.lastValue.onComplete(task)

            assertError(IllegalStateException::class.java)

            dispose()
        }
    }

    @JvmName("verifyAuthResultAddOnCompleteListenerCalled")
    private inline fun Task<AuthResult>.verifyAddOnCompleteListenerCalled() {
        verify(this, times(1))
                .addOnCompleteListener(onAuthResultCompleteListener.capture())
    }

    @JvmName("verifyGetTokenResultAddOnCompleteListenerCalled")
    private inline fun Task<GetTokenResult>.verifyAddOnCompleteListenerCalled() {
        verify(this, times(1))
                .addOnCompleteListener(onGetTokenResultCompleteListener.capture())
    }

    @JvmName("verifyVoidAddOnCompleteListenerCalled")
    private inline fun Task<Void>.verifyAddOnCompleteListenerCalled() {
        verify(this, times(1))
                .addOnCompleteListener(onVoidCompleteListener.capture())
    }
}